### ISSUE TITLE
Unscented Kalman Filter transform : Cannot cast ufunc 'add' output from dtype('float64') to dtype('int32') with casting rule 'same_kind'

### ISSUE THREAD

**Issue by apiszcz** [dstl/Stone-Soup#195]
I am trying to use UKF in place of EKF in the code fragment below. However for UKF i am getting the following stack trace (see below). I noted the defaults are set for UKF if no arguments are given. Am Am I missing setting up something for UKF?

```
from stonesoup.predictor.kalman import UnscentedKalmanPredictor
predictor = UnscentedKalmanPredictor(transition_model)
.
.
.
from stonesoup.updater.kalman import UnscentedKalmanUpdater
updater = UnscentedKalmanUpdater(measurement_model)


tracks = {Track([prior_one]), Track([prior_two])}
for n, (measurements, clutter_set) in enumerate(zip(measurementss, clutter), 1):
    detections = clutter_set.copy()
    detections.update(measurements) # Add measurements and clutter together
    
    hypotheses = data_associator.associate(tracks, detections, start_time+timedelta(seconds=n))
    for track in tracks:
        hypothesis = hypotheses[track]
        if hypothesis.measurement:
            post = updater.update(hypothesis)
            track.append(post)
        else: # When data associator says no detections are good enough, we'll keep the prediction
            track.append(hypothesis.prediction)
```

```
---------------------------------------------------------------------------
UFuncTypeError                            Traceback (most recent call last)
<ipython-input-23-75a62285695b> in <module>
      5     detections.update(measurements) # Add measurements and clutter together
      6 
----> 7     hypotheses = data_associator.associate(tracks, detections, start_time+timedelta(seconds=n))
      8     for track in tracks:
      9         hypothesis = hypotheses[track]

C:\g\mot\lib\a\lib\site-packages\stonesoup\dataassociator\neighbour.py in associate(self, tracks, detections, time)
    100         hypotheses = {
    101             track: self.hypothesiser.hypothesise(track, detections, time)
--> 102             for track in tracks}
    103 
    104         # Link hypotheses into a set of joint_hypotheses and evaluate

C:\g\mot\lib\a\lib\site-packages\stonesoup\dataassociator\neighbour.py in <dictcomp>(.0)
    100         hypotheses = {
    101             track: self.hypothesiser.hypothesise(track, detections, time)
--> 102             for track in tracks}
    103 
    104         # Link hypotheses into a set of joint_hypotheses and evaluate

C:\g\mot\lib\a\lib\site-packages\stonesoup\hypothesiser\distance.py in hypothesise(self, track, detections, timestamp)
     62 
     63         # Common state & measurement prediction
---> 64         prediction = self.predictor.predict(track.state, timestamp=timestamp)
     65         measurement_prediction = self.updater.get_measurement_prediction(
     66             prediction)

C:\g\mot\lib\a\lib\site-packages\stonesoup\predictor\kalman.py in predict(self, prior, control_input, timestamp, **kwargs)
    358             transition_noise_covar, control_input.state_vector,
    359             control_matrix, contol_noise_covar,
--> 360             self.alpha, self.beta, self.kappa)
    361 
    362         return GaussianStatePrediction(prediction_mean,

C:\g\mot\lib\a\lib\site-packages\stonesoup\predictor\kalman.py in predict_lowlevel(x, P, f, Q, u, B, Qu, alpha, beta, kappa)
    402 
    403         sigma_points, mean_weights, covar_weights = \
--> 404             gauss2sigma(x, P, alpha, beta, kappa)
    405 
    406         x_pred, P_pred, _, _, _, _ = unscented_transform(sigma_points,

C:\g\mot\lib\a\lib\site-packages\stonesoup\functions.py in gauss2sigma(mean, covar, alpha, beta, kappa)
    123     # Calculate sigma point locations
    124     sigma_points = np.tile(mean, (1, 2 * ndim_state + 1))
--> 125     sigma_points[:, 1:(ndim_state + 1)] += sqrt_sigma * np.sqrt(c)
    126     sigma_points[:, (ndim_state + 1):] -= sqrt_sigma * np.sqrt(c)
    127 

UFuncTypeError: Cannot cast ufunc 'add' output from dtype('float64') to dtype('int32') with casting rule 'same_kind'
```

**Comment by jmbarr** [comment]
It looks like the defaults which create `c` in `gauss2sigma()` are floats and these may be causing problems when combined with ints in the state vector. I don't know if duck-typing takes care of this in most cases, but there could be occasions where the precision of the float is incompatible with the int. Two options:

quick, less satisfactory: try changing the type in the state vector to float. That might make the problem go away in this instance
slow, better: we adjust `gauss2sigma()` to do sensible casting in the += and -= operations.

**Comment by apiszcz** [comment]
Thank you for the quick reply, the state appears to be in float format.


```
{Clutter(state_vector=StateVector([[ 1.66301316],
              [16.37871774]]), timestamp=datetime.datetime(2020, 4, 17, 6, 25, 28, 432434), measurement_model=None, metadata={}),
 Clutter(state_vector=StateVector([[0.23380983],
              [3.82244584]]), timestamp=datetime.datetime(2020, 4, 17, 6, 25, 28, 432434), measurement_model=None, metadata={}),
 Clutter(state_vector=StateVector([[10.82974423],
              [ 7.39599691]]), timestamp=datetime.datetime(2020, 4, 17, 6, 25, 28, 432434), measurement_model=None, metadata={}),
 Clutter(state_vector=StateVector([[18.06186083],
              [19.81801825]]), timestamp=datetime.datetime(2020, 4, 17, 6, 25, 28, 432434), measurement_model=None, metadata={}),
 Clutter(state_vector=StateVector([[7.15560454],
              [2.75202862]]), timestamp=datetime.datetime(2020, 4, 17, 6, 25, 28, 432434), measurement_model=None, metadata={}),
 Detection(state_vector=StateVector([[Bearing(1.934889777427264)],
              [22.02436236790193]], dtype=object), timestamp=datetime.datetime(2020, 4, 17, 6, 25, 28, 432434), measurement_model=None, metadata={'trkid': 1}),
 Detection(state_vector=StateVector([[Bearing(3.066721073961606)],
              [7.7439653554149075]], dtype=object), timestamp=datetime.datetime(2020, 4, 17, 6, 25, 28, 432434), measurement_model=None, metadata={'trkid': 0})}
```

**Comment by sdhiscocks** [comment]
I've replicated this by passing in state_vector with integers. In the original notebooks, `prior_one` and `prior_two` states are integers. Does changing these to floats fix this @apiszcz?

Seems the in place addition/subtraction breaks this, as trying to modify existing number (and associated place in memory) versus replacing the value with a new number. The following fixes it (if not a little more ugly :wink: ):

```diff
diff --git a/stonesoup/functions.py b/stonesoup/functions.py
index db7567f..0f0ddfc 100644
--- a/stonesoup/functions.py
+++ b/stonesoup/functions.py
@@ -130,8 +130,10 @@ def gauss2sigma(state, alpha=1.0, beta=2.0, kappa=None):
 
     # Calculate sigma point locations
     sigma_points = np.tile(state.state_vector, (1, 2 * ndim_state + 1))
-    sigma_points[:, 1:(ndim_state + 1)] += sqrt_sigma * np.sqrt(c)
-    sigma_points[:, (ndim_state + 1):] -= sqrt_sigma * np.sqrt(c)
+    sigma_points[:, 1:(ndim_state + 1)] = \
+        sigma_points[:, 1:(ndim_state + 1)] + sqrt_sigma*np.sqrt(c)
+    sigma_points[:, (ndim_state + 1):] = \
+        sigma_points[:, (ndim_state + 1):] - sqrt_sigma * np.sqrt(c)
 
     # Put these sigma points into s State object list
     sigma_points_states = []

```

Seems sensible enough fix to me. What are your thoughts @jmbarr?

**Comment by apiszcz** [comment]
Initial test was to change the state_vector for prior_one and prior_two values to floats.

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-19-75a62285695b> in <module>
      5     detections.update(measurements) # Add measurements and clutter together
      6 
----> 7     hypotheses = data_associator.associate(tracks, detections, start_time+timedelta(seconds=n))
      8     for track in tracks:
      9         hypothesis = hypotheses[track]

C:\g\mot\lib\a\lib\site-packages\stonesoup\dataassociator\neighbour.py in associate(self, tracks, detections, time)
    100         hypotheses = {
    101             track: self.hypothesiser.hypothesise(track, detections, time)
--> 102             for track in tracks}
    103 
    104         # Link hypotheses into a set of joint_hypotheses and evaluate

C:\g\mot\lib\a\lib\site-packages\stonesoup\dataassociator\neighbour.py in <dictcomp>(.0)
    100         hypotheses = {
    101             track: self.hypothesiser.hypothesise(track, detections, time)
--> 102             for track in tracks}
    103 
    104         # Link hypotheses into a set of joint_hypotheses and evaluate

C:\g\mot\lib\a\lib\site-packages\stonesoup\hypothesiser\distance.py in hypothesise(self, track, detections, timestamp)
     84             measurement_prediction = self.updater.get_measurement_prediction(
     85                 prediction, detection.measurement_model)
---> 86             distance = self.measure(measurement_prediction, detection)
     87 
     88             # True detection hypothesis

C:\g\mot\lib\a\lib\site-packages\stonesoup\measures.py in __call__(self, state1, state2)
    167             cov = state1.covar
    168 
--> 169         vi = np.linalg.inv(cov)
    170 
    171         return distance.mahalanobis(u, v, vi)

<__array_function__ internals> in inv(*args, **kwargs)

C:\g\mot\lib\a\lib\site-packages\numpy\linalg\linalg.py in inv(a)
    545     signature = 'D->D' if isComplexType(t) else 'd->d'
    546     extobj = get_linalg_error_extobj(_raise_linalgerror_singular)
--> 547     ainv = _umath_linalg.inv(a, signature=signature, extobj=extobj)
    548     return wrap(ainv.astype(result_t, copy=False))
    549 

TypeError: No loop matching the specified signature and casting was found for ufunc inv
```

**Comment by apiszcz** [comment]
Next test is with the proposed code change and prior_one and prior_two state vectors as integers.

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-38-75a62285695b> in <module>
      5     detections.update(measurements) # Add measurements and clutter together
      6 
----> 7     hypotheses = data_associator.associate(tracks, detections, start_time+timedelta(seconds=n))
      8     for track in tracks:
      9         hypothesis = hypotheses[track]

C:\g\mot\lib\a\lib\site-packages\stonesoup\dataassociator\neighbour.py in associate(self, tracks, detections, time)
    100         hypotheses = {
    101             track: self.hypothesiser.hypothesise(track, detections, time)
--> 102             for track in tracks}
    103 
    104         # Link hypotheses into a set of joint_hypotheses and evaluate

C:\g\mot\lib\a\lib\site-packages\stonesoup\dataassociator\neighbour.py in <dictcomp>(.0)
    100         hypotheses = {
    101             track: self.hypothesiser.hypothesise(track, detections, time)
--> 102             for track in tracks}
    103 
    104         # Link hypotheses into a set of joint_hypotheses and evaluate

C:\g\mot\lib\a\lib\site-packages\stonesoup\hypothesiser\distance.py in hypothesise(self, track, detections, timestamp)
     84             measurement_prediction = self.updater.get_measurement_prediction(
     85                 prediction, detection.measurement_model)
---> 86             distance = self.measure(measurement_prediction, detection)
     87 
     88             # True detection hypothesis

C:\g\mot\lib\a\lib\site-packages\stonesoup\measures.py in __call__(self, state1, state2)
    167             cov = state1.covar
    168 
--> 169         vi = np.linalg.inv(cov)
    170 
    171         return distance.mahalanobis(u, v, vi)

<__array_function__ internals> in inv(*args, **kwargs)

C:\g\mot\lib\a\lib\site-packages\numpy\linalg\linalg.py in inv(a)
    545     signature = 'D->D' if isComplexType(t) else 'd->d'
    546     extobj = get_linalg_error_extobj(_raise_linalgerror_singular)
--> 547     ainv = _umath_linalg.inv(a, signature=signature, extobj=extobj)
    548     return wrap(ainv.astype(result_t, copy=False))
    549 

TypeError: No loop matching the specified signature and casting was found for ufunc inv
```

**Comment by jmbarr** [comment]
Wait - where did `predict_lowlevel()` come from? Wasn't that deprecated a few updates ago?

**Comment by apiszcz** [comment]
I don't know the answer on that one.
FYSA: I performed a pip install stonesoup before this test.

**Comment by jmbarr** [comment]
> I've replicated this by passing in state_vector with integers. In the original notebooks, `prior_one` and `prior_two` states are integers. Does changing these to floats fix this @apiszcz?
> 
> Seems the in place addition/subtraction breaks this, as trying to modify existing number (and associated place in memory) versus replacing the value with a new number. The following fixes it (if not a little more ugly 😉 ):
> 
> ```diff
> diff --git a/stonesoup/functions.py b/stonesoup/functions.py
> index db7567f..0f0ddfc 100644
> --- a/stonesoup/functions.py
> +++ b/stonesoup/functions.py
> @@ -130,8 +130,10 @@ def gauss2sigma(state, alpha=1.0, beta=2.0, kappa=None):
>  
>      # Calculate sigma point locations
>      sigma_points = np.tile(state.state_vector, (1, 2 * ndim_state + 1))
> -    sigma_points[:, 1:(ndim_state + 1)] += sqrt_sigma * np.sqrt(c)
> -    sigma_points[:, (ndim_state + 1):] -= sqrt_sigma * np.sqrt(c)
> +    sigma_points[:, 1:(ndim_state + 1)] = \
> +        sigma_points[:, 1:(ndim_state + 1)] + sqrt_sigma*np.sqrt(c)
> +    sigma_points[:, (ndim_state + 1):] = \
> +        sigma_points[:, (ndim_state + 1):] - sqrt_sigma * np.sqrt(c)
>  
>      # Put these sigma points into s State object list
>      sigma_points_states = []
> ```
> 
> Seems sensible enough fix to me. What are your thoughts @jmbarr?

So if I understand this right, the += (and -=) operator work directly on the same place in memory, and if the right hand side of the operation can't 'fit' into the left hand side then an error is chucked. Efficient, but limiting, I guess. 

If the fix works, then great, ugliness notwithstanding (it actually looks more like Matlab). Might be worth adding a note to prevent bright sparks proposing to change the `a = a + b` to `a += b`.

**Comment by sdhiscocks** [comment]
@apiszcz : the new error your seeing occurs in version on PyPI but not on latest _master_, so I assume whatever the issue is we must have fixed. :smile:

I'd recommend installing from the GitHub repo as significant changes and updates have happened. You can do this directly from pip:
```
pip install --force-reinstall --no-deps https://github.com/Dstl/Stone-Soup/archive/master.zip
```

One thing to note with the newer version for the notebooks is the model `RangeBearingGaussianToCartesian` was renamed to `CartesianToBearingRange`.

We'll aim to get version on PyPI updated soon.

**Comment by apiszcz** [comment]
Issues resolved.

1. I created an alias for the pip update, I see stonesoup improving rapidly. Update to latest as instructed.
2. I bumped into the RangeBearingGaussianToCartesian, changed to CartesianToBearingRange.
3. The MTT_UKF notebook i created derived from samples and help here works now. 

I appreciate the rapid response.
I am looking forward creating the particle filter tracking example next, the onto the radar model simulators.

Thank you.
